### ISSUE TITLE
DNS challenge consistently fails for gcloud

### ISSUE THREAD

**Issue by knmorgan** [terraform-providers/terraform-provider-acme#51]
I consistently get the following when trying a DNS challenge with gcloud.

```
* googlecloud: no public zone found for domain sub.example.com.
```

There is a public zone and everything is set up properly with gcloud. I've done quite a bit of testing on various versions of this provider, and am confident this change happened when updating to v1.2.0. It works as expected on v1.1.2.

**Comment by vancluever** [comment]
Hey @knmorgan, only thing that I can think of is there might be some issue with the gcloud provider. Lego was also updated to 2.5.0 in the latest release.

I did mention some issues that I saw in testing the multi-provider update in #50, but I think that might not be 100% related to your issue.

What I would probably do is set up a test case and debug it using [delve](https://github.com/go-delve/delve/). I have a shortcut for running delve through the acceptance test framework:

```
make debugacc TEST=./acme TESTARGS=-run=TestAccACMECertificate_YOURTESTHERE
```

You could then set a breakpoint somewhere in the DNS challenge process and try and figure out what may be happening.

The error message itself seems to indicate that the provider is having trouble discovering the correct public zone name to set up the TXT records in, so that's where I'd start.

PS: You can't use delve itself within Terraform since 1) Terraform itself forks and 2) our [plugin system](https://github.com/hashicorp/go-plugin/) works by spawning providers as external processes and then communicating with them over gRPC.

Good luck!

**Comment by tl-aiyor** [comment]
I have got different issues with gcloud DNS challenge.  For the same credential file that worked a couple of weeks ago, it now gives the following error,

```
* acme_certificate.certificate: 1 error(s) occurred:

* acme_certificate.certificate: googlecloud: unable to get Google Cloud client: google: could not find default credentials. See https://developers.google.com/accounts/docs/application-default-credentials for more information.

Terraform does not automatically rollback in the face of errors.
Instead, your Terraform state file has been partially updated with
any resources that successfully completed. Please address the error
above and apply again to incrementally change your infrastructure.
```
The actual configuration block is,

```
resource "acme_certificate" "certificate" {
  account_key_pem         = "${acme_registration.reg.account_key_pem}"
  certificate_request_pem = "${tls_cert_request.csr.cert_request_pem}"
  min_days_remaining      = 7

  dns_challenge {
    provider = "gcloud"

    config {
      GCE_PROJECT              = "this-project"
      GCE_POLLING_INTERVAL     = "10"
      GCE_PROPAGATION_TIMEOUT  = "180"
      GCE_TTL                  = "60"
      GCE_SERVICE_ACCOUNT_FILE = "${file("./dns_gcp_cred.json")}"
    }
  }
}
```

**Comment by vancluever** [comment]
@tl-aiyor, I have not been able to reproduce this with 1.3.0 unfortunately. Using a test config I was able to effect this successfully by defining `GCE_PROJECT` and `GCE_SERVICE_ACCOUNT_FILE` in the `config` hash.

**Comment by ryansmith23** [comment]
Regarding the initial issue by @knmorgan, I believe the `lego` client had issues with locating gcloud ManagedZones that it fixed in v2.6.0. View the changelog: https://github.com/go-acme/lego/blob/v2.6.0/CHANGELOG.md

I believe with gcloud API you need to include the trailing period to filter by DNS Name when finding ManageZones, but lego didn't allow common names with trailing periods as input.

**Comment by sjmiller609** [comment]
In my case, I am using Google Domains + Google Cloud DNS. I needed to configure the DNS servers in Google Domains to match the NS records in Google Cloud DNS. https://issuetracker.google.com/issues/133640275

**Comment by ryansmith23** [comment]
@sjmiller609 what are you passing for `common_name` to your `acme_certificate`? I was using `${google_dns_managed_zone.<name>.dns_name}` but this was failing. This domain looked like `example.com.`. When I used this it failed to find the managed zone for the domain. If I just manually passed in `example.com` I was getting some other error.

I am also using Google Domains and have my DNS setup properly and currently use the domain, just can't get the certificate registered.

**Comment by yellowmegaman** [comment]
Same here, everything works with TF 0.12 + acme 1.3.4 for route53, but not for gcloud

**Comment by sjmiller609** [comment]
@ryansmith23
https://github.com/astronomer/terraform-google-astronomer-gcp/blob/master/domain.tf

**Comment by vchrisb** [comment]
I have the same issue:
```
* googlecloud: no public zone found for domain sub.example.com.
* error encountered while presenting token for DNS challenge: googlecloud: no public zone found for domain sub.example.com.
```

**Comment by yellowmegaman** [comment]
My problem was:
```
Error: googlecloud: project name missing
```

After enabling `TF_LOG: debug` I found the main problem with unset env variables, which are available here: https://github.com/go-acme/lego/blob/master/providers/dns/gcloud/googlecloud.go

So anyone stumbling on same thing, just set:
`GCE_SERVICE_ACCOUNT` and `GCE_PROJECT` in addition to terraform env vars.

Previously it was working without them. Since for terraform to function properly it's only needed to set `GOOGLE_CREDENTIALS` + `GOOGLE_PROJECT` + `GOOGLE_REGION`

**Comment by roidelapluie** [comment]
https://github.com/go-acme/lego/pull/875

**Comment by roidelapluie** [comment]
@vancluever can we get a LEGO update? 
go-acme/lego#875 seems to be the fix

**Comment by renchap** [comment]
lego update is #79

**Comment by vancluever** [comment]
Hey all. This will be coming this maintenance cycle, but I was waiting to see what the outcome of #79 was going to be before I committed to pulling in the next batch of lego updates. I'll probably have time to work on this this weekend.

**Comment by lorainekv** [comment]
Hi there, wanted to check in and see what the release schedule looks like for this? Excited to see all the progress that has been made!

**Comment by vancluever** [comment]
@lorainekv hopefully the next release will be in the next week. There is one more PR to merge before we cut it. Stay tuned!

**Comment by vancluever** [comment]
Should be fixed with the v1.4.0 release.
